src/metabase/driver/sql/query_processor.clj

```diff
@@ -107,6 +124,27 @@
 (defmethod add-interval-honeysql-form :sql [driver hsql-form amount unit]
   (driver/date-add driver hsql-form amount unit))
 
+(defn adjust-start-of-week
+  "Truncate to the day the week starts on."
+  [driver truncate-fn expr]
+  (let [offset (driver.common/start-of-week-offset driver)]
+    (if (not= offset 0)
+      (add-interval-honeysql-form driver
+                                  (truncate-fn (add-interval-honeysql-form driver expr offset :day))
+                                  (- offset) :day)
+      (truncate-fn expr))))
+
+(defn adjust-day-of-week
```

Comment:
I didn't test, but I think this will break in the client. We need to remap the offset numbers to the correct labels based on the setting.